File: frontend/apps/ios/Pantopus/Features/Ballot/BallotGlyph.swift
//
//  BallotGlyph.swift
//  Pantopus
//
//  The canvas's ballot-box glyph (Board: Place: Your ballot card) and
//  the 34pt home-green tile that carries it. Not a Lucide icon: the
//  canvas draws its own three strokes on a 24-unit grid, reproduced
//  here exactly (2-unit stroke, round caps and joins).
//

import SwiftUI

struct BallotGlyphShape: Shape {
    func path(in rect: CGRect) -> Path {
        let s = min(rect.width, rect.height) / 24
        func p(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: rect.minX + x * s, y: rect.minY + y * s)
        }
        var path = Path()
        // M4 11h16v9a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1z — the box.
        path.move(to: p(4, 11))
        path.addLine(to: p(20, 11))
        path.addLine(to: p(20, 20))
        path.addArc(tangent1End: p(20, 21), tangent2End: p(19, 21), radius: s)
        path.addLine(to: p(5, 21))
        path.addArc(tangent1End: p(4, 21), tangent2End: p(4, 20), radius: s)
        path.closeSubpath()
        // M9 11V4h6v7 — the ballot sheet.
        path.move(to: p(9, 11))
        path.addLine(to: p(9, 4))
        path.addLine(to: p(15, 4))
        path.addLine(to: p(15, 11))
        // M8 16h8 — the slot.
        path.move(to: p(8, 16))
        path.addLine(to: p(16, 16))
        return path
    }
}

struct BallotGlyph: View {
    var size: CGFloat = 19
    var color: Color = Theme.Color.home

    var body: some View {
        BallotGlyphShape()
            .stroke(color, style: StrokeStyle(lineWidth: 2 * size / 24, lineCap: .round, lineJoin: .round))
            .frame(width: size, height: size)
            .accessibilityHidden(true)
    }
}

/// 34pt rounded tile, home-green, radius 9 — the Place section tile.
struct BallotTile: View {
    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 9, style: .continuous)
                .fill(Theme.Color.homeBg)
            BallotGlyph(size: 19)
        }
        .frame(width: 34, height: 34)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/ballot/BallotGlyph.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.ballot

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.scale
import androidx.compose.ui.graphics.vector.PathParser
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.pantopus.android.ui.theme.PantopusColors

/**
 * The canvas's ballot-box glyph (Board: Place: Your ballot card): the box,
 * the ballot sheet and the slot, on a 24-unit grid. Not a Lucide icon —
 * the canvas draws its own three strokes, reproduced here from its path
 * data (2-unit stroke, round caps and joins). Parity twin of iOS
 * `BallotGlyph.swift`.
 */
private const val GLYPH_PATH = "M4 11h16v9a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1z M9 11V4h6v7 M8 16h8"
private const val GLYPH_GRID = 24f

@Composable
fun BallotGlyph(
    modifier: Modifier = Modifier,
    size: Dp = 19.dp,
    color: Color = PantopusColors.home,
) {
    val path = remember { PathParser().parsePathString(GLYPH_PATH).toPath() }
    Canvas(modifier = modifier.size(size)) {
        scale(scale = this.size.minDimension / GLYPH_GRID, pivot = Offset.Zero) {
            drawPath(path, color, style = Stroke(width = 2f, cap = StrokeCap.Round, join = StrokeJoin.Round))
        }
    }
}

/** 34dp rounded tile, home-green, radius 9 — the Place section tile. */
@Composable
fun BallotTile(modifier: Modifier = Modifier) {
    Box(
        modifier = modifier.size(34.dp).clip(RoundedCornerShape(9.dp)).background(PantopusColors.homeBg),
        contentAlignment = Alignment.Center,
    ) {
        BallotGlyph()
    }
}
